### PR TITLE
Rewrite Gemma4 scannable block

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

Currently the implementation of Gemma4ScannableBlock makes XLA treat the whole 6-layer block as a "big layer". The all-gathers for all 6 layers are run at the beginning of each block. Similarly, in backward pass, remat runs the whole block. This causes elevated HBM usage and more exposed collectives. This PR builds on top of my previous PRs #4506 and #4507, which allows a more modular and clean implementation.

After this rewrite:
* 5 local layers via jax.lax.scan, each per-layer jax.checkpoint
* length-1 scan for global layer, also wrapped in jax.checkpoint
* remove block-level remat and unroll outer block scan

### Added after revision:
* Refactor in gemma4.py
* Fix ckpt conversion: update checkpoint conversion code for both to_maxtext and to_hf directions.
* Fix gemma4-26b: support sowing of `moe_lb_loss` from layer intermediates to fix a bug breaking 26b model
* Fix decode: add "local_layer" to logical rules to fix a bug breaking vllm decode introduced in #4066

# Tests
* added unit test
* Before rewrite: [xprof](http://xprof.corp.google.com/trace_viewer/aireenmei-16525735383985599335) for pdbs=4. [xprof](http://xprof.corp.google.com/trace_viewer/aireenmei-3752206657021917776) for pdbs=8
* After rewrite: [xprof](https://xprof.corp.google.com/trace_viewer/aireenmei-2068683819829996003?view_start=-6192.735&view_end=47760.970) for pdbs=4. [xprof](http://xprof/?session_id=aireenmei-3228254048924162039) for pdbs=8 + vocab_tiling=8 (OOM without vocab_tiling). 
* Before rewrite, the HBM peak is not on vocab peak([screenshot](https://screenshot.googleplex.com/BNLXydTTJqgsFGp)). After rewrite, vocab peak becomes the highest ([screenshot](https://screenshot.googleplex.com/oPgh9cAzZs7E6oW)) due to other peaks become smaller, also by having remat boundary at layer level adds HBM usage and overlap with the vocab peak. Therefore, without vocab tiling, the peak HBM usage doesn't change much compared with original. But when combined with vocab_tiling, we can effectively lower the HBM peak a lot (70 GB -> 40 GB for pdbs=4).

### Added after revision
* checkpoint conversion to_maxtext & logit check:
  * [31B log](https://paste.googleplex.com/6723341921681408)
  * [26B log](https://paste.googleplex.com/4505618714066944)
  * added unit test in parem_mapping_test.py
* checkpoint conversion to_hf:
  * [31B to_hf log](https://paste.googleplex.com/5038062890516480) 
* Decode:
  * [26B decode script](https://paste.googleplex.com/6634462606655488) similar to the test script in #4066
  * [26B decode log](https://paste.googleplex.com/6541973380005888)
  * added uinit test to guard the kv_cache support

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by aireenmei** [comment]
I add pull ready to generate internal cl for testing. Won't submit until PR approved

File: src/maxtext/models/gemma4.py [hunk 1]
```diff
@@ -429,27 +433,110 @@ def __init__(
...
     self.rngs = rngs
     self.num_of_layers = num_of_layers
+    self.remat_policy_fn = remat_policy_fn
+    self.apply_internal_remat = apply_internal_remat
+
+    pattern_length = len(GEMMA4_ATTENTION_PATTERN)
+    if not 0 <= num_of_layers <= pattern_length:
+      raise ValueError(f"Gemma4ScannableBlock must contain between 0 and {pattern_length} layers; got {num_of_layers}.")
+
+    # Pattern is 5 local, 1 global.
```
Comment (RissyRan) [inline src/maxtext/models/gemma4.py]: It seems the pattern has been defined. Will something like this work?

```
    active_pattern = GEMMA4_ATTENTION_PATTERN[:num_of_layers]
    self.num_local = sum(1 for t in active_pattern if t == AttentionType.LOCAL_SLIDING)
    self.num_global = sum(1 for t in active_pattern if t == AttentionType.GLOBAL)
```

File: src/maxtext/models/gemma4.py [hunk 2]
```diff
@@ -429,27 +433,110 @@ def __init__(
...
+              rngs=layer_rngs,
+              attention_type=AttentionType.LOCAL_SLIDING,
+              layer_idx=0,  # layer_idx is not used in the class
+          ),
+          length=self.num_local,
+          param_scan_axis=self.config.param_scan_axis,
+          metadata_axis_name="local_layers",
+          rngs=self.rngs,
+      )
+    else:
```
Comment (RissyRan) [inline src/maxtext/models/gemma4.py:486]: Will this be triggered for Gemma4?

File: src/maxtext/models/gemma4.py [hunk 3]
```diff
@@ -471,22 +558,120 @@ def __call__(
     y = inputs
 
     updated_kvs = []
-    for layer_id in range(self.num_of_layers):
-      current_kv = kv_cache[layer_id] if kv_cache is not None else None
-      y, new_kv = getattr(self, f"layers_{layer_id}")(
+
+    if kv_cache is not None:
```
Comment (RissyRan) [inline src/maxtext/models/gemma4.py:681]: I guess you meet issue when calling `kv_cache[layer_id]`? 

Wondering if you tested for decoding with the change end-to-end and see reasonable outputs?

File: src/maxtext/models/gemma4.py [hunk 4]
```diff
@@ -471,22 +558,120 @@ def __call__(
...
+          updated_kvs.append(new_kv)
+          per_layer_states.append(nnx.state(layer))
+
+        stacked_state = jax.tree.map(lambda *xs: jnp.stack(xs), *per_layer_states)
+        if scan_axis != 0:
+          stacked_params, stacked_other = stacked_state.split(nnx.Param, ...)
+          stacked_params = jax.tree.map(lambda x: jnp.moveaxis(x, 0, scan_axis), stacked_params)
+          stacked_state = nnx.State.merge(stacked_params, stacked_other)
+        nnx.update(self.local_layers, stacked_state)
+    elif self.local_layers is not None:
```
Comment (RissyRan) [inline src/maxtext/models/gemma4.py]: Is this the same as line 564?

**Review by RissyRan — COMMENTED** [review]
Thanks for the change! Overall it looks good. I am a little bit lost in gemma4.py if/else branches, wondering if we could simplify a little bit?

File: src/maxtext/layers/nnx_scan.py [hunk 5]
```diff
@@ -87,3 +88,50 @@ def update_leaf(leaf):
...
+    carry: Any,
+    *,
+    length: int,
+    param_scan_axis: int,
+    apply_fn: Callable[[nnx.Module, Any], Any],
+    remat: bool = False,
+    remat_policy: Callable[..., Any] | None = None,
+    prevent_cse: bool = True,
+    unroll: int = 1,
+) -> Any:
```
Comment (NuojCheng) [inline src/maxtext/layers/nnx_scan.py:104]: is this funtion only used in gemma4? I imagine every models gonna use this function, for my own understanding, why this code block is created in a gemma4-related PR?

File: src/maxtext/models/gemma4.py [hunk 6]
```diff
@@ -429,27 +433,110 @@ def __init__(
...
+      )
+      return layer_out[0] if isinstance(layer_out, tuple) else layer_out
+
+    apply_remat = self.apply_internal_remat and self.config.remat_policy != "none"
+    if apply_remat:
+      prevent_cse = maxtext_utils.should_prevent_cse_in_remat(self.config)
+      remat_policy = self.remat_policy_fn
+    else:
+      prevent_cse = True
+      remat_policy = None
```
Comment (NuojCheng) [inline src/maxtext/models/gemma4.py]: for my own understanding, why fix prevent cse as true when apply remat is false?

File: src/maxtext/models/gemma4.py [hunk 7]
```diff
@@ -471,22 +558,120 @@ def __call__(
...
+        # working set and OOMs. Offloaded (pinned-host) residuals can't cross that
+        # boundary, so the global layer saves would-be-offloaded tensors on device
+        # instead; the local-layer scan (a real multi-iteration scan) still offloads.
+        global_remat_policy = self.remat_policy_fn
+        offload_names = maxtext_utils.get_save_and_offload_names(cfg)
+        if offload_names[0] or offload_names[1]:
+          save_names, offload_to_device = offload_names
+          global_remat_policy = jax.checkpoint_policies.save_only_these_names(*(save_names + offload_to_device))
+
+        if self.apply_internal_remat and self.config.remat_policy != "none":
```
Comment (NuojCheng) [inline src/maxtext/models/gemma4.py]: config.remat_policy is never "none"? When it can be "none"?

File: src/maxtext/models/gemma4.py [hunk 8]
```diff
@@ -471,22 +558,120 @@ def __call__(
...
+          scan_global_layer = jax.checkpoint(
+              scan_global_layer,
+              policy=global_remat_policy,
+              prevent_cse=prevent_cse,
+          )
+
+        # Carry state through the loop instead of returning a stacked [1, ...]
+        # scan result: slicing that result previously introduced a bitcast
+        # between device and pinned-host memory under offload remat.
+        with xla_metadata.set_xla_metadata(**{"skip-simplify-while-loops_trip-count-one": "true"}):
```
Comment (NuojCheng) [inline src/maxtext/models/gemma4.py]: I am thinking we could move this context manager to train.py with a maxtext flag controlling it, wdyt? @gobbleturk @RissyRan

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/models/gemma4.py (hunk 8 above)
Comment (NuojCheng) [reply src/maxtext/models/gemma4.py]: in a separate PR of course, let's keep it in this PR

File: src/maxtext/models/gemma4.py (hunk 1 above)
Comment (aireenmei) [reply src/maxtext/models/gemma4.py]: code updated as suggested

File: src/maxtext/models/gemma4.py (hunk 2 above)
Comment (aireenmei) [reply src/maxtext/models/gemma4.py:486]: Added comments line 466-469
```
    # num_of_layers can be 0: the decoders always construct a "remainder" block
    # for num_decoder_layers % pattern_length layers, which is 0 whenever the
    # layer count divides evenly (e.g. 31b = 60 layers). That block is built but
    # never applied, so num_local or num_global may legitimately be 0 here.
```

File: src/maxtext/models/gemma4.py (hunk 3 above)
Comment (aireenmei) [reply src/maxtext/models/gemma4.py:681]: The kv_cache related changes are to preserve the feature introduced in #4066. I added unit tests "test_gemma4_block_external_kv_cache_matches_scanned_path" to guard this feature for future. Also attached end to end decode results in PR description test section

File: src/maxtext/models/gemma4.py (hunk 4 above)
Comment (aireenmei) [reply src/maxtext/models/gemma4.py]: The original code checks "local_layers is not None" in both the kv_cache branch and the non-kv cache branch, causing confusion. This section has now been refactored with helper functions for each branch ([screenshot](https://screenshot-v2.corp.google.com/A3rwUymc3WrxEeF)). Hopefully it's easier to follow

File: src/maxtext/layers/nnx_scan.py (hunk 5 above)
Comment (aireenmei) [reply src/maxtext/layers/nnx_scan.py:104]: Other models have been using [NNXDecoder._apply_layers_sequentially](https://github.com/AI-Hypercomputer/maxtext/blob/6d89c4176b8785d60501b829911ad345f6eccb81/src/maxtext/layers/nnx_decoders.py#L865). I added cross reference in the docstring of both functions.
```
    This is a minimal, model-agnostic scan primitive. The other NNX decoder
    paths instead go through ``NNXDecoder._apply_layers_sequentially``, a heavier
    applier that also threads external (vLLM) KV caches via a static unroll and
    re-applies scan-axis metadata. Gemma4 is currently the only caller of this
    function; unifying the two appliers is a follow-up cleanup.
```

File: src/maxtext/models/gemma4.py (hunk 6 above)
Comment (aireenmei) [reply src/maxtext/models/gemma4.py]: Thanks for catching. prevent_cse is actually a no-op when apply_remat is False, so the value doesn't matter. Now removed "prevent_cse = True", will leave it the default value.

File: src/maxtext/models/gemma4.py (hunk 7 above)
Comment (aireenmei) [reply src/maxtext/models/gemma4.py]: decoders.get_remat_policy checks `if cfg.remat_policy != "none"` (https://github.com/AI-Hypercomputer/maxtext/blob/df95a1fcdccc78166bc453faf016c680351eeefb/src/maxtext/layers/decoders.py#L349). Looks like "none" is allowed, at least historically. We can deprecate it later if no longer needed. But for now, I will keep it to align with the current code.

File: src/maxtext/models/gemma4.py (hunk 8 above)
Comment (aireenmei) [reply src/maxtext/models/gemma4.py]: sounds good

**Review by RissyRan — APPROVED** [review]
